clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -294,18 +365,33 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
         if (error != Errors.NONE && canRetry(batch, error)) {
             // retry
             log.warn("Got error produce response with correlation id {} on topic-partition {}, retrying ({} attempts left). Error: {}",
-                     correlationId,
-                     batch.topicPartition,
-                     this.retries - batch.attempts - 1,
-                     error);
-            this.accumulator.reenqueue(batch, now);
-            this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+                    correlationId,
+                    batch.topicPartition,
+                    this.retries - batch.attempts - 1,
+                    error);
+            if (transactionState == null || (transactionState.pid() == batch.producerId())) {
+                // If idempotence is enabled only retry the request if the current PID is the same as the pid of the
+                // batch.
+                this.accumulator.reenqueue(batch, now);
+                this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+            }
```

Comment:
Don't we need to do something in the `else` case? Something like what's done below, completing the batch using `done` and deallocating it.